channeldb/mp_payment.go

```diff
@@ -359,6 +348,86 @@ func (m *MPPayment) UpdateState() error {
 	return m.updateState()
 }
 
+// NeedWaitAttempts decides whether we need to hold creating more HTLC attempts
+// and wait for the results of the payment's inflight HTLCs. Return an error if
+// the payment is in an unexpected state.
+func (m *MPPayment) NeedWaitAttempts() (bool, error) {
+	// Check when the remainingAmt is not zero, which means we have more
+	// money to be sent.
+	if m.State.RemainingAmt != 0 {
+		switch m.Status {
+		// If the payment is newly created, no need to wait for HTLC
+		// results.
+		case StatusInitiated:
+			return false, nil
+
+		// If we have inflight HTLCs, we'll check if we have terminal
+		// states to decide if we need to wait.
+		case StatusInFlight:
+			// We still have money to send, yet one of the HTLCs is
+			// settled, return an error as the peer is violating
+			// the protocol.
+			if m.State.HasSettledHTLC {
+				return false, fmt.Errorf("%w: settled HTLC "+
+					"but still have remaining amount %v",
+					ErrPaymentInternal,
+					m.State.RemainingAmt)
+			}
+
+			// Otherwise we don't need to wait for inflight HTLCs
+			// since we still have money to be sent.
+			return false, nil
+
+		// We need to send more money, yet the payment is already
+		// succeeded. Return an error in this case as the peer is
+		// violating the protocol.
+		case StatusSucceeded:
```

Comment:
Q: How can we end up in this state, where we think the payment is succeeded but still trying to send something?